src/transport/MessageEndpointServer.cpp

```diff
@@ -11,150 +13,300 @@ namespace faabric::transport {
+    // https://github.com/zeromq/cppzmq/issues/478
+    if (syncFanIn != nullptr) {
+        syncFanIn->stop();
+    }
+
+    if (asyncFanIn != nullptr) {
+        asyncFanIn->stop();
+    }
+
+    // Join each worker
+    for (auto& t : workerThreads) {
+        if (t.joinable()) {
+            t.join();
+        }
+    }
+
+    // Join the receiver thread
+    if (receiverThread.joinable()) {
+        receiverThread.join();
     }
 }
 
-MessageEndpointServer::MessageEndpointServer(int asyncPortIn, int syncPortIn)
+MessageEndpointServer::MessageEndpointServer(int asyncPortIn,
+                                             int syncPortIn,
+                                             const std::string& inprocLabelIn,
+                                             int nThreadsIn)
   : asyncPort(asyncPortIn)
   , syncPort(syncPortIn)
-  , asyncThread(this, true)
-  , syncThread(this, false)
+  , inprocLabel(inprocLabelIn)
+  , nThreads(nThreadsIn)
+  , asyncHandler(this, true, inprocLabel + "-async", nThreadsIn)
+  , syncHandler(this, false, inprocLabel, nThreadsIn)
   , asyncShutdownSender(LOCALHOST, asyncPort)
   , syncShutdownSender(LOCALHOST, syncPort)
 {}
 
+/**
+ * We need to guarantee to callers of this function, that when it returns, the
+ * server will be ready to use.
+ */
 void MessageEndpointServer::start()
 {
-    // This latch means that callers can guarantee that when this function
-    // completes, both sockets will have been opened (and hence the server is
-    // ready to use).
+    // This latch allows use to block on the handlers until _just_ before they
+    // start their proxies.
     auto startLatch = faabric::util::Latch::create(3);
 
-    asyncThread.start(startLatch);
-    syncThread.start(startLatch);
+    asyncHandler.start(startLatch);
+    syncHandler.start(startLatch);
 
     startLatch->wait();
+
+    // Unfortunately we can't know precisely when the proxies have started,
+    // hence have to add a sleep.
+    SLEEP_MS(500);
 }
 
 void MessageEndpointServer::stop()
 {
-    // Send shutdown messages
-    SPDLOG_TRACE(
-      "Server sending shutdown messages to ports {} {}", asyncPort, syncPort);
+    // Here we send shutdown messages to each worker in turn, however, because
+    // they're all connected on the same inproc port, we have to wait until each
+    // one has shut down fully (i.e. the zmq socket has gone out of scope),
+    // before sending the next shutdown message (hence the use of the latch). If
+    // we don't do this, zmq will direct messages to sockets that are in the
+    // process of shutting down and cause errors.
+    for (int i = 0; i < nThreads; i++) {
+        SPDLOG_TRACE("Sending async shutdown message {}/{} to port {}",
+                     i + 1,
+                     nThreads,
+                     asyncPort);
+
+        setShutdownLatch();
+        asyncShutdownSender.send(shutdownHeader.data(), shutdownHeader.size());
+        awaitShutdownLatch();
```

Comment:
It is counter-intuitive that we use this method here, and not in line 165. I imagine it is because `awaitShutdownLatch` nulls out the pointer, but still makes it hard to debug and follow the code.